### PR TITLE
ci: Testnet deployment automation

### LINKED ISSUE #147: Bug: cannot rerun build
seems like the worker is not created second time when you rerun build with exactly same commit, this is caused most likely with the clash of hash. this could be solved by naming workers by random hash generated in 1st step and propagated through outputs.

### PR DESCRIPTION
**Pull request by fakirAyoub**
## Description
Automate the deployment of k8s components. On push to master:
1. Run build on AWS
2. Build docker image
3. Publish it to ECR
4. Dispatch to K8s deployment repository

## Motivation and Context
Automating the creation of testnet end-to-end after each release

## How Has This Been Tested?
End to end testing cross repositories
Check K8S deployment

fixes #147

### CODE REVIEW

File: .github/workflows/workflow.yml
Comment (maht0rz) [inline .github/workflows/workflow.yml:118]: don't think you'd have to use SUDO on a CI runner, might be potentionally unsafe.

File: .github/workflows/workflow.yml
Comment (maht0rz) [inline .github/workflows/workflow.yml:115]: can we parametrise all these URLs?

File: .github/workflows/workflow.yml
Comment (maht0rz) [inline .github/workflows/workflow.yml:252]: can we get a notification on successful deployment as well, not only build?

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:118]: Agree but it's a temporary machine so

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:115]: Yes!

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:252]: Yup. Adding that. Will resolve conversation once done

Comment (fakirAyoub) [reply .github/workflows/workflow.yml:252]: Added in the testnet-deployment repo

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:226]: push it to the dockerhub & github instead, i recommend using action to simplify it like here https://github.com/galacticcouncil/HydraDX-node/blob/master/.github/workflows/docker.yml#L77

**Review by lumir-mrkva — CHANGES_REQUESTED** [review]

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:238]: Thanks @lumir-mrkva , will make the change

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:256]: tag the images based on the branch, `master -> latest`, `pr-branch -> parsed-branchname/commit` hash

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:275]: run deployment only when pipeline is ran in the master branch

File: docker/Dockerfile
Comment (lumir-mrkva) [inline docker/Dockerfile:1]: move this file to the / please

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:257]: @lumir-mrkva can you advise the syntax you have in mind?

File: docker/Dockerfile
Comment (fakirAyoub) [reply docker/Dockerfile]: Sure, done

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:256]: major problem is taht you cannot use `/` in tag so you have to replace it

so for example if the build is from branch `feat/awesome-feature` it would be tagged as `pr-feat-awesome-feature`

later we can also trigger build on git tags and in that case it would tag it without parsing like `v11.0.1` for example

Comment (fakirAyoub) [reply .github/workflows/workflow.yml:256]: Okay, so, our main build would be the one tagged as `master`. For the branches, I will replace `/` with - before naming.

Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:256]: master branch should be tagged as `latest`, as that is default tag that docker tries to pull

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:146]: you don't have to wait for `upload-config-to-s3` as they're not running on that runner

```suggestion
    needs: build-and-publish-docker-image
```

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:110]: rename it to `basilisk-testnet-chainspec.json`

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:198]: why sleep?

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:213]: why sleep?

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:215]: docker should be already in ami

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:225]: @fakirAyoub this is still an issue

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:223]: @fakirAyoub still not resolved

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [inline .github/workflows/workflow.yml:197]: why is this removed?

File: Dockerfile
Comment (lumir-mrkva) [inline Dockerfile:5]:
```suggestion
WORKDIR /basilisk
```


note that you'll have to change following command accordingly (remove basilisk from path)

File: Dockerfile
Comment (lumir-mrkva) [inline Dockerfile:19]: ENTRYPOINT, but i prefer CMD instead if possible

File: Dockerfile
Comment (lumir-mrkva) [inline Dockerfile:1]: use slimmer image

**Review by lumir-mrkva — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (lumir-mrkva) [reply Dockerfile:19]: note that with entrypoint you cannot use basilisk, you can only execute testing basilisk

File: Dockerfile
Comment (lumir-mrkva) [inline Dockerfile:13]: image should be able to run both basilisk and testing basilisk, eg. basilisk should be copied and then testing basilisk hard linked

File: Dockerfile
Comment (lumir-mrkva) [reply Dockerfile:1]: disregard, ubuntu is using slim images by default

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:146]: We do, because the file is uploaded from this machine, when it start destroying we cannot get the file from it anymore

File: Dockerfile
Comment (martinfridrich) [reply Dockerfile:19]: Why do you prefer CMD? You can override both. Entrypoint is easier to be used by para registration script. If you change it to CMD, don't forget to update para. registration script.

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:215]: for some reason, without doing this install step, I have an error saying that Docker cannot be found.

Weird error since the build jobs use it correctly... Can we address this in another PR?

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:197]: was for test only, adding it before merge

File: Dockerfile
Comment (fakirAyoub) [reply Dockerfile:19]: let's keep ENTRYPOINT for now

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:215]: i just thought it is already there as i am using it here https://github.com/galacticcouncil/HydraDX-node/blob/master/.github/workflows/docker.yml#L46

Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:215]: mby those actions before ensure that docker is installed already? or the ami is setup differently?

File: Dockerfile
Comment (lumir-mrkva) [reply Dockerfile:19]: ok

File: .github/workflows/workflow.yml
Comment (fakirAyoub) [reply .github/workflows/workflow.yml:215]: I have to check the AMI for Basilisk

File: .github/workflows/workflow.yml
Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:146]: thats not true, you are uploading it from github ubuntu, see https://github.com/galacticcouncil/Basilisk-node/pull/263/files/8ba96bc1635496a9240397f0b28524889d67b934#diff-126bf89616b7daa3d14ebc882ad18666aaf1c3dae888c4ba306a66ec80758bc1R126

Comment (lumir-mrkva) [reply .github/workflows/workflow.yml:146]: the file is already uploaded to github artifacts even before build job finishes

File: Dockerfile
Comment (lumir-mrkva) [reply Dockerfile:19]: @martinfridrich can the normal basilisk be the default entrypoint then at least ? :)

Comment (martinfridrich) [reply Dockerfile:19]: @lumir-mrkva sure, we can make special case for basilisk in para reg. srcipt. Basilisk will be better entrypoint if this images will be public. I just wanted to have unified way to call subcmds in docker images by script. This change have to  be reflected into para reg. script

**Review by lumir-mrkva — APPROVED** [review]

**Comment by lumir-mrkva** [comment]
@martinfridrich please merge this when rest of scripts are ready and the docker file default entrypoint changed

**Comment by lumir-mrkva** [comment]
fixes #147

**Pull request merged**
